### ISSUE TITLE
counsel "find-X" commands worse at default guessing than the corresponding builtins

### ISSUE THREAD

**Issue by stepnem** [abo-abo/swiper#1088]
I would expect the counsel replacements to be at least as smart as the builtin commands they purport to supersede, but they're not (in this respect at least).

E.g., the builtin `find-function` guesses the right default anywhere inside the function call. Counsel had better make use of the preexisting and well-tried builtin guessing functions, e.g. the patch below seems to fix the `find-function` case. There are equivalents for `find-variable`, `find-library`...

```diff
diff --git a/counsel.el b/counsel.el
index 6bd25a7..3ec8c6e 100644
--- a/counsel.el
+++ b/counsel.el
@@ -524,7 +524,8 @@ (defun counsel-describe-function ()
                      (push (symbol-name x) cands))))
                 cands)
               :keymap counsel-describe-map
-              :preselect (ivy-thing-at-point)
+              :preselect (when-let (fn (function-called-at-point))
+                           (symbol-name fn))
               :history 'counsel-describe-symbol-history
               :require-match t
               :sort t
```

**Comment by abo-abo** [comment]
I disagree with the proposed change, it's better the way it is now.

There's value of looking up specifically the symbol at point. There's also value in seeing it's not bound. You can always press <kbd>C-M-u</kbd> if you want to describe the containing function.

**Comment by stepnem** [comment]
Well, of course the cases you mention are covered by the builtins,
too, unless I'm missing something. I guess the fact I messed up find-
and describe- in the OP didn't help, but it's related.

A concrete example:

In the form

  (setq ivy-on-del-error-function 'ignore)

With point on "ignore" or "setq", both describe-function and
counsel-describe-function will do the (same) right thing.

With point somewhere else, e.g. on "ivy-on-del-error-function",
describe-function still suggests "setq" as the function (macro) being
called at point, while counsel-describe-function doesn't suggest
anything.

I've found that behaviour utterly frustrating. You really think it's
better? I hope not, but if you do, what about counsel-find-library,
which fails to even suggest the name of loaded feature at point (e.g.
with point at "ivy" in "(use-package ivy)"). Again, find-library does
the right thing.
